### ISSUE TITLE
http caches do not detect session

### ISSUE THREAD

**Issue by bbalban** [expressjs/session#588]
Hello,
I have a standard ExpressJs application with html pages using express-session. Express-session works perfectly fine without caching. When I serve pages through a cache, the webcache does not distinguish whether there is a session or not.  (Tried this on Cloudfront and Fastly, based on Varnish)

Is there a solution for this? 

The behavior I expect is that the web cache detects a session and forwards requests to the origin server when there is a session, and serve from the cache when there is no session.

Since caching is common this would be a great use case to add to docs IMO.

Thanks!
Bahadir

**Comment by bbalban** [comment]
@dougwilson do you mean you never faced this situation, or you did but never had the problem?

perhaps the cache can only know a valid session by asking the server or inspecting the cookie.

**Comment by dougwilson** [comment]
Oh, sorry, i mean that I have never tried to cache dynamic content before, so never had to figure out how to do that properly.

**Comment by wesleytodd** [comment]
This issue is something you will have to figure out yourself because it depends on your providers and configurations.  When I have done this in the past I have used the presence of the session cookie or the auth header.  If the cookie is not being removed you can add some code to unset the cookie in your logout route.

FWIW, I actually usually try to avoid using a cdn cache for content which is dynamic by user.  It can introduce enough complexity when debugging that it is not worth it IMO.  So if you have servers which can handle the load, do it on your own.

**Comment by bbalban** [comment]
Thanks for the insight. I think the solution would be to reliably update the cookie such that the cache service can recognize a valid and invalid session on later requests.

I added this to the logout route: 

```
// logout users
exports.logout = function(req, res) {
	req.logout();
	req.session.destroy();
+	res.clearCookie('connect.sid', {path: '/'});
	res.redirect('/');
}
```
I see below in the response headers: 
```
Set-Cookie: connect.sid=; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 GMT
```

However on subsequent GET requests I still see the original cookie:
```
connect.sid=s%3Adt61yHTO79ZkbXutnWGopikMvXG77HbL.GW3yQdY5m7AnPADD%2BIRAUCKcO4k8ZlvOi4%2FlKrnBJg8
```
Could it be because I do a redirect? It shouldn't matter I suppose. I'm using google chrome on a Mac.

**Comment by knoxcard** [comment]
Aren't you always issued a session ID?  Is that the same session ID or a new one?

**Comment by bbalban** [comment]
It is the same one. You still have a session ID after logout? I thought below line should clear it.
`res.clearCookie('connect.sid', {path: '/'});`

**Comment by bbalban** [comment]
You are right, I got issued a new session ID. I didn't know a new SID is issued after logout.

**Comment by bbalban** [comment]
That makes it difficult for the cache to distinguish a regular visitor and one with session, if a SID is issued for every visitor.

**Comment by dougwilson** [comment]
You can configure it many ways. In order to help, we need you to provide access to your source code so we can understand what is happening. The README has all the options documented if that help, but if we can see the source we can provide extra guidance.

**Comment by bbalban** [comment]
Hello, thanks for all the help. I came to realize connect.sid can be set even though there is no authenticated session. It makes sense that session does not necessarily imply authentication.

I am sharing my solution for reference in case anyone needs it though it does not involve express-session.

*Login*
I added another cookie on every login, in Passport.js Strategy:

Set ```    passReqToCallback: true,```
Then in succesful authentication within strategy function body:
```req.res.cookie("auth_state", "probably_loggedin");```

*Logout*
```res.clearCookie('auth_state');```

Then in my case the Fastly CDN is configured as follows: https://community.fastly.com/t/need-guidance-on-caching-behavior-with-regard-to-session-cookies/1301?u=bbalban

Thanks for all the help.

Bahadir
